Fix stuck navigation & Re-add Contact Method Details pages

aldo-expensify opened the pull request:
### Details

Re-implemented all of the changes from this PR (https://github.com/Expensify/App/pull/15204) which was reverted, but this PR fixes the regression and adds back the old changes

To fix [the bug](https://github.com/Expensify/App/pull/15751#issuecomment-1467068484), a functionality to keep track when navigations transitions are in progress was removed: https://github.com/Expensify/App/pull/15751/commits/b659e1bbca2d03002f4b01cc96f716ce96b31d7a . That code was originally fixing this issue https://github.com/Expensify/App/pull/14426, so we should check that we are not introducing the double tapping bug again.

<details>
<summary>TL;DR about the bug</summary>


1. There is a double navigation happening:
    - Navigate to `ROUTES.SETTINGS_CONTACT_METHODS` when we confirm deleting the phone number in the modal ([here](https://github.com/Expensify/App/blob/99a10acf856f27ec5b799aeb3da3943348e263cc/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js#L144))
    - Inside the `render` function we navigate to `ROUTES.SETTINGS_CONTACT_METHODS` if the login data is not found ([here](https://github.com/Expensify/App/blob/6b2a7690ee170994f1e533473b88fb4802957b54/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js#L142-L145)). The login data is not found because the [successData](https://github.com/Expensify/App/blob/99a10acf856f27ec5b799aeb3da3943348e263cc/src/libs/actions/User.js#L243-L249) of the request removes it from Onyx. 
1. `ScreenWrapper` reacts to the `'transitionStart'` event setting `Navigation.setIsNavigating(true);` ([here](https://github.com/Expensify/App/blob/4237497d7e803acbe52533d12e0b46c6071ee4b6/src/components/ScreenWrapper/index.js#L41-L43))
3. `ScreenWrapper` calls `Navigation.setIsNavigating(false);` on `'transitionEnd'`, but this never happens because we [unsubscribe to these events](https://github.com/Expensify/App/blob/4237497d7e803acbe52533d12e0b46c6071ee4b6/src/components/ScreenWrapper/index.js#L85-L90) before getting the `'transitionEnd'` event.

</details>

### Fixed Issues

$ https://github.com/Expensify/App/issues/15725
$ https://github.com/Expensify/App/issues/15203
$ https://github.com/Expensify/App/issues/11575

### Tests

Along with the below steps, also perform:
- the tests found in https://github.com/Expensify/App/pull/15204, and
- the tests found in https://github.com/Expensify/App/pull/14426

Connected to the staging API (couldn't reproduce connected to the dev API):

1. Login in NewDot and go to `/settings/profile/contact-methods`
4. Add a phone number (no need to validate)
5. You should be taken back to the "Contact methods" page:
    ![image](https://user-images.githubusercontent.com/87341702/223825442-58c1e9df-ca23-4179-a589-120f941893f0.png)
6. Click on your phone number contact method and then delete it
    ![image](https://user-images.githubusercontent.com/87341702/223825619-c3e20f79-56f8-46e1-a654-d9c2c0bd4388.png)
7. Verify that you are taken back to the "Contact methods" page
8. Verify that you can tap in any of the menu entries and you get navigated
![image](https://user-images.githubusercontent.com/87341702/223840508-57e8c100-9b5d-415a-8a78-48fa023f8608.png)

- [x] Verify that no errors appear in the JS console

### Offline tests

1. Login in NewDot and go to `/settings/profile/contact-methods`
3. Add a phone number (no need to validate)
4. You should be taken back to the "Contact methods" page:
    ![image](https://user-images.githubusercontent.com/87341702/223825442-58c1e9df-ca23-4179-a589-120f941893f0.png)
5. Click on your phone number contact method
6. GO OFFLINE
7. Click the delete button and confirm in the modal
9. Verify that you are taken back to the "Contact methods" page
10. Verify that the deleted contact method is still there with strike-through


### QA Steps

Along with the below steps, also perform:
- the tests found in https://github.com/Expensify/App/pull/15204, and
- the tests found in https://github.com/Expensify/App/pull/14426

1. Login in NewDot and go to `/settings/profile/contact-methods`
2. Add a phone number (no need to validate)
3. You should be taken back to the "Contact methods" page:
    ![image](https://user-images.githubusercontent.com/87341702/223825442-58c1e9df-ca23-4179-a589-120f941893f0.png)
4. Click on your phone number contact method and then delete it
    ![image](https://user-images.githubusercontent.com/87341702/223825619-c3e20f79-56f8-46e1-a654-d9c2c0bd4388.png)
5. Verify that you are taken back to the "Contact methods" page
9. Verify that you can tap in any of the menu entries and you get navigated

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3885503/221638255-cb4a221b-2b79-4782-9b59-c330f9ca4fe5.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3885503/222133669-a621b291-cb81-46cd-95eb-f961298ab0ad.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3885503/222138546-304da173-e438-48b1-b8cc-c5be7261a266.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3885503/221642089-6a9f0311-9135-4fde-b091-4868b8f9a255.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3885503/222137317-dc5b927d-c604-4964-a6fc-7d97259fc4dc.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3885503/222132844-7d75dbb6-a866-445c-8eaf-e3c8373d7f48.mov

</details>

aldo-expensify committed:
No need to call Navigate.navigate

aldo-expensify commented:
cc @roryabraham @parasharrajat @syedsaroshfarrukhdot In case you have a better understanding of these navigation transitions and can say if this makes sense or not 🙏

aldo-expensify commented:
Making it a draft again because I noticed that this has an unwanted effect: if you are offline, `successData` doesn't remove the contact method from Onyx and you are not navigated back until you come back offline.

Without this PR, when offline, we land here:

![image](https://user-images.githubusercontent.com/87341702/223829213-9d9a0e66-7595-4c35-a5a9-7542ac524b8b.png)


With this PR, when offline, we land here:

![image](https://user-images.githubusercontent.com/87341702/223829830-b328ae4f-3a4d-4f22-bbfc-b2d39c73dd6e.png)

aldo-expensify commented:
The code navigating automatically to the main contacts methods page when the contact method is not found is here:

https://github.com/Expensify/App/blob/4237497d7e803acbe52533d12e0b46c6071ee4b6/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js#L146-L149

So the problem seems to be cause by this double navigation, I see two options for now:

- Don't navigate if the contact method is not found, but show some page mentioning that the information was not found, or
- Try to fix ScreenWrapper and Navigation.isNavigating stuff so it doesn't get stuck if we navigate twice.

aldo-expensify committed:
Redirect on componentDidUpdate instead of render

aldo-expensify commented:
Update:
- We now redirect in `componentDidUpdate` if the login data is not found (instead of in `render()`)
- I added a [loading indicator](https://github.com/Expensify/App/blob/99a10acf856f27ec5b799aeb3da3943348e263cc/src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js#L165) in `render` if the login information is not available.

This seems to fix the bug correctly without making the offline experience worse.

aldo-expensify commented:
> Which one is the correct behavior?

The one on top (edited the comment to make it more clear)



> If ScreenWrapper is getting unmounted before we can set the flags then we should move the logic of setting `Navigation.setIsNavigating(value)` to the parent component which is never mounted.
> 
> I haven't validated the root cause but I believe ya 😉
> 
> We can move this setup to top-level navigators in `AuthScreens.js` & `PublicScreens.js`.

Oh, that sounds like an interest approach, I can give it a try.

Besides that, maybe we should avoid navigating during rendering anyway. What do you think about that? I feel like the rendering function should be exclusively about rendering (showing data) and not triggering side-effects.

aldo-expensify commented:
@parasharrajat about this suggestion https://github.com/Expensify/App/pull/15751#issuecomment-1460799962

The difficulty I see is that `ScreenWrapper` has this [state.didScreenTransitionEnd](https://github.com/Expensify/App/blob/fb2014493988d2f3d30f1bcae160c1d307b5da7e/src/components/ScreenWrapper/index.js#L52) that changes on `'transitionEnd'`.

If we moved the event listening part to higher in the tree to `AuthScreens.js` & `PublicScreens.js`, I'm not sure how would `ScreenWrapper` get notified when the transition ends. `ScreenWrapper` is used in many places, so drilling a state like that from `AuthScreens.js` & `PublicScreens.js` until it gets to `ScreenWrapper` sounds like a looot of work. The other option would be to keep this state in Onyx so we can access it easily in `ScreenWrapper`, but I'm not sure if there could be some bad consequence of using Onyx as storage here. Do you have some thoughts on that?

aldo-expensify commented:
The PR adding the ContactMehodDetailsPage got reverted here: https://github.com/Expensify/App/pull/15761

Beamanator commented:
This seems to work quite well, I don't think I've ever noticed the loading indicator, have you @aldo-expensify ?

Beamanator committed:
Add everything back that got reverted

Beamanator commented:
FYI I re-added pretty much everything from my old PR into this one: https://github.com/Expensify/App/pull/15777 just to see if I could reproduce the bug and I haven't been able to, would either of y'all (@aldo-expensify @parasharrajat ) mind trying to reproduce on that PR?

aldo-expensify commented:
> FYI I re-added pretty much everything from my old PR into this one: #15777 just to see if I could reproduce the bug and I haven't been able to, would either of y'all (@aldo-expensify @parasharrajat ) mind trying to reproduce on that PR?

I'll give it a try

aldo-expensify commented:
> I think we can keep the transitionEnd event handler in screenWrapper component which is responsible for didTransitionEnd. I am only thinking to move the isNavigaying flag setter to high level. Thus both level the AuthScreens and screenWrapper will have event listener.

I see, that should be much easier, I think we should do that.

Beamanator committed:
Fix lint and missing LoginField

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into aldo_fix-delete-phone-number

Beamanator committed:
Merge branch 'beaman-recreateShowAllExpensifyLogins' into aldo_fix-delete-phone-number

Beamanator commented:
@aldo-expensify hope you don't mind, but I merged `main` to get up to date with removing all the related pages (from the revert PR https://github.com/Expensify/App/pull/15761), and adding everything back in https://github.com/Expensify/App/pull/15777.

Then I closed that PR and moved those changes here so we can test out your fix(es). I'll also update the test steps to mention the original test steps from the original reverted PR (https://github.com/Expensify/App/pull/15039)

Beamanator committed:
A bit of cleanup

neil-marcellini commented:
Yeesh this looks _involved_ haha. Please tag me when it's ready for review.

aldo-expensify commented:
Agreed with @parasharrajat about the moving the `isNavigating` stuff, lets do it in a follow up.

aldo-expensify commented:
> @aldo-expensify hope you don't mind, but I merged `main` to get up to date with removing all the related pages (from the revert PR #15761), and adding everything back in #15777.

I don't mind, I'm just surprised that you did this. I would have thought that you would get a ton of conflicts 🤮

Beamanator commented:
Haha ya @aldo-expensify the only conflicts where the changes you made in the ContactMethodDetailsPage, but those were pretty easy to resolve.

@aldo-expensify I think we can mark this as "Ready for review", then maybe I can redo all of the old tests (all the new pages, green indicators, red dot indicators, offline UX) and would you mind doing videos / screenshots for the bug fix flow? Just b/c I was never able to reproduce that haha

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_fix-delete-phone-number

aldo-expensify commented:
@Beamanator reproduced the original bug:


https://user-images.githubusercontent.com/87341702/224834538-4338b830-4ebe-4aff-b6ff-4c8a10d60728.mov

aldo-expensify commented:
> @aldo-expensify Shnikeys :/ can you explain in detail how you can reproduce?? 🙏

Same as before:

1. Add unverified phone number
2. Delete it
3. Try to click one of the menu entries -> nothing happens

https://user-images.githubusercontent.com/87341702/224847514-bf11d77a-675d-4125-952c-a6104a6af39a.mov

aldo-expensify committed:
Merge with main / resolve conflicts

aldo-expensify commented:
@parasharrajat @Beamanator 

I have spent a few hours trying [this](https://github.com/Expensify/App/pull/15751#issuecomment-1460799962) and I haven't been able to make it work. You can play around with this branch if you want to try it yourself: https://github.com/Expensify/App/tree/aldo_trying-to-move-listener-uo

The problem I encounter is that any listener higher in the tree doesn't get reliably the `transitionEnd` and `transitionStart` events, and I'm guessing this is because there isn't a single Navigator ([RootStack.Navigator](https://github.com/Expensify/App/blob/9b97d0b4a4299f8b2a8485c40e1d36fe509055e4/src/libs/Navigation/AppNavigator/AuthScreens.js#L177), [MainDrawerNavigator](https://github.com/Expensify/App/blob/9b97d0b4a4299f8b2a8485c40e1d36fe509055e4/src/libs/Navigation/AppNavigator/AuthScreens.js#L216-L219), [ModalStackNavigators....Navigator](https://github.com/Expensify/App/blob/9b97d0b4a4299f8b2a8485c40e1d36fe509055e4/src/libs/Navigation/AppNavigator/ModalStackNavigators.js)), there are multiple nested and I think the event stops at the first one.

Considering that relocating the code is not as easy as it looks, I started to question why do we even have this listener that sets [Navigation.isNavigating](https://github.com/Expensify/App/blob/9b97d0b4a4299f8b2a8485c40e1d36fe509055e4/src/libs/Navigation/Navigation.js#L32). I understand that this functionality was added by this https://github.com/Expensify/App/pull/14426 to fix https://github.com/Expensify/App/issues/14258 (double tapping the currency can potentially navigate you back twice)

I tried to reproduce the issue https://github.com/Expensify/App/issues/14258 disabling the `Navigation.isNavigating` stuff and I couldn't, it is either hard to reproduce or it is not happening any more because of other changes.

Considering all that, I think the best would be to revert https://github.com/Expensify/App/pull/14426:
- I think the possibility of leaving someone with the app completely locked is very bad ([this bug](https://github.com/Expensify/App/pull/15751#issuecomment-1467068484)), and the gains don't justify it. I think it would have been safer to just prevent the double tap at the level of the input with some flag changing value.
- The original issue it was solving doesn't seem to happen anymore.

cc @roryabraham since you may have more context about https://github.com/Expensify/App/pull/14426

Agree?

Beamanator commented:
Wow thanks for the great investigation @aldo-expensify ! I will try to find some time to talk to @roryabraham about this today since we're luckily in the same location for this week 👍

Beamanator commented:
@aldo-expensify I'm going to try to reproduce this morning - and if I can, I am going to consider trying to use Navigation.canNavigate in ContactMethodDetailsPage to check if we can navigate or not - I wonder if you've tried this to avoid the double-navigate-stuck state?

aldo-expensify commented:
I'm going to proceed with removing the `Navigation.isNavigating` stuff (more details: https://github.com/Expensify/App/pull/15751#issuecomment-1467192602) to unblock the work on the contact methods pages.

aldo-expensify committed:
Remove Navigation.isNavigating

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -0,0 +1,272 @@
+
+const defaultProps = {
+    loginList: {},
+    route: {
+        params: {
+            contactMethod: '',
+        },
```
The problem I see with that is then you get this error:

<img width="839" alt="image" src="https://user-images.githubusercontent.com/87341702/225739824-51f41282-ce53-4f4b-9624-33f3943ec9e8.png">

And I'm not sure that making `route/params/contactMethod` non required is the best, considering that it is required?

Beamanator replied:
Aah yeah that's interesting, I am not sure it's beneficial in this situation to use `initialParams` - but this could be an interesting discussion if there's some drawback of the current solution? Otherwise I think we should keep what we have as @aldo-expensify said

aldo-expensify commented on src/libs/actions/User.js:
```diff
@@ -100,6 +101,61 @@ function resendValidateCode(login) {
+ */
+function requestContactMethodValidateCode(contactMethod) {
+    const optimisticData = [{
+        onyxMethod: CONST.ONYX.METHOD.MERGE,
+        key: ONYXKEYS.LOGIN_LIST,
```
hmm, I honestly prefer the current approach. Using a variable like `onyxData` saves reduces very little the code repetition at the cost of hiding a bit the `onyxMethod` and `key` used.

aldo-expensify commented:
@Beamanator  @neil-marcellini @parasharrajat this is ready for review

cc @cristipaval in case you want to review / have a look

Beamanator commented:
Oh ok @aldo-expensify so this will basically just be a straight revert-revert?

aldo-expensify commented:
> Oh ok @aldo-expensify so this will basically just be a straight revert-revert?

The PR you mean? 

It is a revert-revert plus:
- fix [bug](https://github.com/Expensify/App/pull/15751#issuecomment-1467068484) by remove the `Navigation.isNavigating` stuff: https://github.com/Expensify/App/pull/15751/commits/b659e1bbca2d03002f4b01cc96f716ce96b31d7a
- move navigation when the login data is not found from `render()` to `componentDidUpdate()` (this part is not important for it to work, but I think it is better to put there from a code responsibility point of view - I don't think we should have redirections/side effects during rendering if possible)

koko57 commented:
For me it looks ok, now it works as expected and doesn't cause any regression.

aldo-expensify committed:
Resolve conflicts / merge main

aldo-expensify commented:
Resolved conflicts

neil-marcellini reviewed (commented):
I found a bug following this flow. Do you agree that this is a problem? If so do you want to fix it here or in a follow up? The same kind of problem occurs on main.

1. Log in with an account that only has an email contact method
2. Go to Settings > Profile > Contact method
3. Click Phone number
4. Enter +17206574277 and click Send validation
5. Verify that you see the error `An error occurred using the phone number provided, please use your email address instead.`
6. Click the arrow in the header to go back to the contact methods page
7. Refresh the page

Expected: The invalid phone number contact method is not added to the page
Actual: The invalid phone number contact method is added to the page and you are asked to verify it.

neil-marcellini reviewed (commented):
Wheew that was a lot. Overall it looks quite good. I tested the main pieces on web. I'm a bit confused about why some of the styling changes were made. I have some comments about the comments too. After that I'll approve.

neil-marcellini commented:
> My take on it is that we can address that in a follow up to keep moving if the problem is already in `main`.

👍 sounds good.



> From point 4., I understand that you added the unverified phone number successfully.

Huh I guess so, but the API command returned an error. I'll provide a video. If I sign out the phone number is still there. Maybe we should not show an error message in this case because clearly it was added, but I guess it failed to be verified? Weird. As a user I have no idea what to do based on the error message.


https://user-images.githubusercontent.com/26260477/226487346-c4a09bec-85fb-4a47-a83b-6ee501773fe7.mov

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -0,0 +1,272 @@
+
+    /** Login list for the user that is signed in */
+    loginList: PropTypes.shape({
+        /** Value of partner name */
```
Lol, I don't know, there is this list here: https://github.com/Expensify/Auth/blob/2f6c3c81afa4dfe4e07b146e3d40c59c4fc6db06/auth/lib/Partner.cpp#L17-L28

neil-marcellini replied:
Let's use something like this? https://github.com/Expensify/Web-Expensify/blob/6cb0434dfaf963520fe92a07b123701edc197867/lib/Auth.php#L591

Or a combination with this https://github.com/Expensify/Web-Expensify/blob/6cb0434dfaf963520fe92a07b123701edc197867/lib/Auth.php#L901

aldo-expensify replied:
Thanks, I'll use those

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -0,0 +1,272 @@
+        if (this.doesContactMethodExist()) {
+            return;
+        }
+        Navigation.navigate(ROUTES.SETTINGS_CONTACT_METHODS);
```
I didn't consider it really, but sounds I wouldn't mind changing to showing a "not found" page... @Beamanator thoughts?

Do we have some guidelines where we say we have this preference?

I see that we recently did this for reports here: https://github.com/Expensify/App/pull/15499

neil-marcellini replied:
I think that's the general pattern that we should follow because anyone can navigate to any URL, and IMO it makes more sense to show that the page is not found vs immediately navigate the user to a different URL.

aldo-expensify replied:
> I think that's the general pattern that we should follow because anyone can navigate to any URL, and IMO it makes more sense to show that the page is not found vs immediately navigate the user to a different URL.

Ok, I'll update it to this

aldo-expensify replied:
![image](https://user-images.githubusercontent.com/87341702/226722096-b63d556c-5a75-4ccf-bb33-942008ce7ade.png)

Updated!

aldo-expensify commented on src/styles/styles.js:
```diff
@@ -1029,31 +1029,18 @@ const styles = {
-        right: 4,
-        bottom: 4,
+        right: -2,
+        top: -1,
```
@Beamanator ?

aldo-expensify replied:
Here is a comparison of how it looked before and after:

<table>
<tr>
<td>
Before
</td>
<td>
After
</td>
</tr>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/87341702/226723658-1757c707-e993-4c58-93ee-548692a2812c.png">
</td>
<td>
<img src="https://user-images.githubusercontent.com/87341702/226723655-e7f408de-fbc9-4653-9a25-b46696e2dca8.png">
</td>
</tr>

aldo-expensify replied:
@neil-marcellini ahh it is intended. Shawn requested these changes here:

https://github.com/Expensify/App/pull/15204#issuecomment-1452333201

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodsPage.js:
```diff
@@ -28,111 +33,136 @@ const propTypes = {
-        if (_.isEmpty(stateToUpdate)) {
-            return;
+        // Temporary checks to determine if we need to show specific LoginField
+        // components. This check will be removed soon.
```
There is a PR @cristipaval is working on that will allow adding multiple phone numbers. The PR is on hold while we get this one merged.

I think @cristipaval 's PR is [this one](https://github.com/Expensify/App/pull/15330), but he or @Beamanator probably know better about this

neil-marcellini replied:
Ok let's put that info in the comment then, that way it will be obvious if we somehow forget. Also the reader will have context about why it's going to be removed.

neil-marcellini replied:
Ok let's put that info in the comment then, that way it will be obvious if we somehow forget. Also the reader will have context about why it's going to be removed.

aldo-expensify replied:
Update comment adding a link to the follow up PR: https://github.com/Expensify/App/pull/15330

aldo-expensify committed:
Polish

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodsPage.js:
```diff
@@ -28,111 +33,136 @@ const propTypes = {
+        // Temporary checks to determine if we need to show specific LoginField
+        // components. This check will be removed soon.
+        // Also we still use login.partnerUserID here even though it could have been
+        // deleted optimistically because if the deletion is pending, do want to show
```
Done

aldo-expensify commented on src/pages/settings/Profile/ProfilePage.js:
```diff
@@ -20,8 +22,21 @@ import * as ReportUtils from '../../../libs/ReportUtils';
+import * as UserUtils from '../../../libs/UserUtils';
 
 const propTypes = {
+    /* Onyx Props */
```
Agreed, @Beamanator probably copy pasted this stuff. I personally don't care, so I agree with NAB and not spending time on this.

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodsPage.js:
```diff
@@ -28,111 +33,136 @@ const propTypes = {
         partnerUserID: PropTypes.string,
 
         /** Date of when login was validated */
+        /** Date login was validated, used to show brickroad info status */
```
Done

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -0,0 +1,272 @@
+            return <FullScreenLoadingIndicator />;
+        }
+
+        const isDefaultContactMethod = (this.props.session.email === loginData.partnerUserID);
```
Removed

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodDetailsPage.js:
```diff
@@ -0,0 +1,272 @@
+     * Toggle delete confirm modal visibility
+     * @param {Boolean} shouldOpen
+     */
+    toggleDeleteModal(shouldOpen) {
```
Updated

neil-marcellini reviewed (commented):
Looking good, some small things from me. Please ping me when it's updated.

aldo-expensify committed:
Show NOT FOUND when contact method is missing

aldo-expensify committed:
Update comments

aldo-expensify committed:
Update comment about follow up PR

aldo-expensify commented:
> Looking good, some small things from me. Please ping me when it's updated.

@neil-marcellini ping 👋

aldo-expensify commented:
@parasharrajat this is ready for review

parasharrajat reviewed (changes requested):
The code looks in good shape.

aldo-expensify commented on src/components/AvatarWithIndicator.js:
```diff
@@ -55,45 +68,50 @@ const defaultProps = {
+    ];
 
     return (
-        <View style={[isLarge ? styles.avatarLarge : styles.sidebarAvatar]}>
```
Hmm, it does seem to be the case, but it looks like we removed it partially:
- `props.size` is still passed down to `<Avatar>` [here](https://github.com/Expensify/App/blob/6e0de6243ae1ac64f623cf39df733e252c1abb34/src/components/AvatarWithIndicator.js#L112)
- `props.size` still defined in `<AvatarWithIndicator>` [here](https://github.com/Expensify/App/blob/6e0de6243ae1ac64f623cf39df733e252c1abb34/src/components/AvatarWithIndicator.js#L26-L27)

I'll remove it completely.

aldo-expensify committed:
Remove unused prop size

aldo-expensify committed:
Merge branch 'main' of github.com:Expensify/App into aldo_fix-delete-phone-number

aldo-expensify committed:
Update comment

Co-authored-by: Rajat Parashar <parasharrajat@users.noreply.github.com>

aldo-expensify commented:
> > My take on it is that we can address that in a follow up to keep moving if the problem is already in `main`.
> 
> 👍 sounds good.
> 
> > From point 4., I understand that you added the unverified phone number successfully.
> 
> Huh I guess so, but the API command returned an error. I'll provide a video. If I sign out the phone number is still there. Maybe we should not show an error message in this case because clearly it was added, but I guess it failed to be verified? Weird. As a user I have no idea what to do based on the error message.

@neil-marcellini This is a problem in the back end, we insert the entry for the secondary login first, and then we try to send the SMS to validate the phone number. It is on this second step that we have this check and throw, then you see the error in the front end, but the phone was added.

https://github.com/Expensify/Web-Expensify/blob/04892cf77cecf32caf75ecc6772b74a1f58f7544/lib/UserAPI.php#L930-L946

I'll move the check to before we add the login to the database. Related convo: https://github.com/Expensify/Web-Expensify/pull/36152/files#r1102989414

neil-marcellini reviewed (dismissed):
I think this is good to go.

aldo-expensify committed:
Fix margins

aldo-expensify committed:
Use CONST

aldo-expensify commented on src/pages/settings/Profile/Contacts/ContactMethodsPage.js:
```diff
@@ -1,138 +1,168 @@
                     <LoginField
-                        label={this.props.translate('common.phoneNumber')}
+                        label={props.translate('common.phoneNumber')}
                         type="phone"
```
Done!

aldo-expensify committed:
Submit validation code on press enter

aldo-expensify committed:
Restrict autocomplete to phones only

aldo-expensify committed:
Remove unused css

aldo-expensify commented:
@parasharrajat 

> BUG: Pressing enter does not submit the verify contact page.

Fixed

> Faced the same issue https://github.com/Expensify/App/pull/15751#pullrequestreview-1349255347. Good to debug in followup.

This will be fixed by a PR in the backend

> BUG: The phone number is showing the wrong autoComplete list(data of username of the app).

Applied your suggestion, it seems to work fine!

> Bug: Inconsistent space between the header and paragraph between the Contact Details page and the default contact method page.

Fixed

Thanks for catching those. Can you have a look again :)

aldo-expensify committed:
Merge main / Resolve conflicts

aldo-expensify commented:
Resolved conflicts.

aldo-expensify commented:
I wonder why there is this error now:

![image](https://user-images.githubusercontent.com/87341702/227322595-06afb0b3-9c85-4bf6-8e43-6e8b5eaf4013.png)


Do we have a new rule enforcing default props for props that are not required?

aldo-expensify commented:
> > Do we have a new rule enforcing default props for props that are not required?
> 
> I think it was there from the start.

but github was not complaining before, it just started now that I merged with `main`

aldo-expensify commented:
also what does this mean:

![image](https://user-images.githubusercontent.com/87341702/227323831-a593ba5a-65ca-4956-b5c2-f830a6cf37da.png)

We can't use "required" anymore?

aldo-expensify committed:
Remove required from prop types / add defaults

aldo-expensify commented:
Updated prop types removing `.isRequired` and added missing defaults

parasharrajat reviewed (changes requested):
Running more tests now.

aldo-expensify commented on src/pages/settings/Profile/Contacts/AddSecondaryLoginPage.js:
```diff
@@ -134,6 +134,7 @@ class AddSecondaryLoginPage extends Component {
                             keyboardType={this.formType === CONST.LOGIN_TYPE.PHONE
                                 ? CONST.KEYBOARD_TYPE.PHONE_PAD : CONST.KEYBOARD_TYPE.EMAIL_ADDRESS}
                             returnKeyType="done"
+                            autoComplete="tel"
```
oops, you are right, I the code wrong and thought this was wrapped in `this.formType === CONST.LOGIN_TYPE.PHONE`, but it was only the label prop

aldo-expensify replied:
Done!

aldo-expensify committed:
Pass correct autoComplete value depending on type

aldo-expensify commented:
> Question: When the user removes a contact offline, Should it be navigatable? ![image](https://user-images.githubusercontent.com/24370807/227380443-fe44e278-5090-42a6-b670-3896fa404696.png)
> 
> It does not open the details page on click.

No, it is disabled.

JmillsExpensify commented:
Great testing and improvements happening in this PR. Thank you!

aldo-expensify commented:
> BUG: After pressing send validations, the browser prompts to save the password. (https://github.com/Expensify/App/pull/15751#issuecomment-1482080346)

We will get rid of the passwords input eventually, and this page will be re-done in a follow up [PR](https://github.com/Expensify/App/pull/15330).

> BUG: Error is not cleared when we navigate again to the add phone number page. (https://github.com/Expensify/App/pull/15751#issuecomment-1482085694)

I noticed this too, but we are deciding not to improve/fix it because we are re-doing this page in the follow up [PR](https://github.com/Expensify/App/pull/15330).

> Bug or not: Is it fine submitting the verify form even if the magic code field is empty?

Good catch, we will fix it

> Bug: clicking or tapping on the empty area trigger resend button.

Good catch, we will fix it

Beamanator committed:
Add new validation to magic code input

Beamanator committed:
Reuse existing validate code validation

Beamanator committed:
Clear error on successful submit

aldo-expensify committed:
Only handle touch in resend validate code text

aldo-expensify commented:
@shawnborton where do you think is better the green check position?

Current:
<img width="381" alt="Screenshot 2023-03-27 at 4 59 03 PM" src="https://user-images.githubusercontent.com/87341702/228013247-a78f57b2-b6a3-4b1f-8ea2-f612ba8b5c0d.png">

Proposed:
<img width="388" alt="Screenshot 2023-03-27 at 4 59 23 PM" src="https://user-images.githubusercontent.com/87341702/228013252-5abe0dd8-05ef-4201-9f83-74fb1caf2f4d.png">

shawnborton commented:
I like the "current" version but I would move it to be a bit closer to the text. Maybe we could try it at 4px margin as well as 8px margin?

Beamanator commented:
Also FYI - the offline behavior for verifying a contact method is a bit confusing to me and @aldo-expensify - previously we've decided it should be pattern B but that doesn't seem super appropriate anymore, and we're not sure how to make that look very nice. So I'm planning to open a follow-up discussion to resolve that - for now I think we can leave move forward w/out completely making that flow perfect (offline verify contact method, pattern B)

aldo-expensify committed:
4px to the left

shawnborton commented:
Ah sorry for the confusion - I meant, can I see it where there is just 4px between the text and the checkbox? And then where there is just 8px between the text and the checkbox? Thanks!

aldo-expensify commented:
> Ah sorry for the confusion - I meant, can I see it where there is just 4px between the text and the checkbox? And then where there is just 8px between the text and the checkbox? Thanks!

16px:
<img width="380" alt="image" src="https://user-images.githubusercontent.com/87341702/228023537-1a52fe32-42e6-42ee-b57e-8a7878c49eaf.png">

12px:
<img width="365" alt="image" src="https://user-images.githubusercontent.com/87341702/228023475-c73d61a3-e161-4693-a1ee-85ca08aafb33.png">

8px:
<img width="376" alt="image" src="https://user-images.githubusercontent.com/87341702/228023641-8db6a83c-64ac-4a74-88da-b36c9e2bcecf.png">

4px:
<img width="365" alt="image" src="https://user-images.githubusercontent.com/87341702/228023713-d5f06464-1665-476e-aea0-ce7989153a5f.png">

Beamanator committed:
Fix lint issues

Beamanator committed:
Make resend checkmark 4 px from text

Beamanator committed:
Merge branch 'main' of github.com:Expensify/App into aldo_fix-delete-phone-number

Beamanator commented:
@parasharrajat I think we're ready for you to test! 👍  I also just made sure the latest build works on iOS 👍  (since it seems you were having trouble a few days ago)

Beamanator commented:
> I think we can just use keyboardShouldPersistTaps='handled' https://reactnative.dev/docs/scrollview#keyboardshouldpersisttaps

Aah that looks nice and simple, but maybe this should be a larger app audit to see which pages don't have that - but this is better as a follow-up

As for [this comment](https://github.com/Expensify/App/pull/15751#issuecomment-1482065302), the button styles were approved so let's keep them as is for now 👍

aldo-expensify committed:
Fix bug requiring double tap to submit

Beamanator commented:
Thanks @aldo-expensify for fixing the android double tap issue

Beamanator commented:
@parasharrajat I'll look into that one real quick, though I'd appreciate if you could start final testing everything else in the meantime 🙏  The blue checkmark button is a very very tiny case out of all the other cases so honestly I don't mind trying to implement that now or later - I really want the contact method details page merged ASAP so @cristipaval can move his PR forward which is holding on this one

Beamanator committed:
All keypad input to blur on submit

parasharrajat reviewed (approved):
There are a few minor bugs that does not affect the main functionality and can be solved in follow-up PR. I think this PR is urgent so let's merge it. 

## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

cc: @Beamanator @neil-marcellini @cristipaval 

:ribbon: :eyes: :ribbon:  C+ reviewed

aldo-expensify commented:
> **Present on Main as well.**
> 
> BUG: Web: Remove Confirm modal position is wrong. ![image](https://user-images.githubusercontent.com/24370807/228334250-5da8e116-871c-49c5-a0e7-d911e2447fce.png)

this one looks very bad, I'll fix it now.

aldo-expensify commented:
> > **Present on Main as well.**
> > BUG: Web: Remove Confirm modal position is wrong. ![image](https://user-images.githubusercontent.com/24370807/228334250-5da8e116-871c-49c5-a0e7-d911e2447fce.png)
> 
> this one looks very bad, I'll fix it now.

nevermind, that is also in staging

Beamanator commented:
@parasharrajat thanks for finding a few follow-up bugs we need to handle! I'd say we don't need to report them in slack, we can handle them in follow-ups in one of the related issues 👍  I'll probably create a new issue for the follow-ups

neil-marcellini reviewed (approved):
NAB, let's tackle these in a follow up, but here are a couple things that I find odd.
- We should definitely center the confirmation modal for removing a contact method
- Why aren't we using the Form component? We should use it for every form and if it doesn't work then we should modify it.
- Which offline UX pattern are we using for adding a phone number? I ask because when I click "Send validation" I have no feedback that anything happened. IMO that's not acceptable.

Let's merge this thing!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4557058308) to staging by https://github.com/neil-marcellini in version: 1.2.92-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/4574189619) to production by https://github.com/luacmartins in version: 1.2.92-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

sobitneupane commented:
[This issue](https://github.com/Expensify/App/issues/19471) was not considered in this PR.

Issue: Contact Method - Validation error message doesn't disappear upon requesting new magic code

ikevin127 commented on src/libs/actions/User.js:
```diff
@@ -187,6 +315,62 @@ function validateLogin(accountID, validateCode) {
+ * @param {String} contactMethod - The contact method the user is trying to verify
+ * @param {String} validateCode
+ */
+function validateSecondaryLogin(contactMethod, validateCode) {
```
Looks like the `validateSecondaryLogin` was missing the `successData` data for the case where, when we have a non-validated account, add a secondary login, and then verify the secondary login, the BE actually sets the secondary login as the primary login.

It can be verified from an email received after verifying, telling that the primary login has been changed to the secondary login. So, when we try to set the secondary login as the primary login, the BE responds with a message that the secondary login is already the primary login.

Leading to this issue:
- https://github.com/Expensify/App/issues/46350

which was fixed by PR:
- https://github.com/Expensify/App/pull/47132

more details on the fix approach can be found in the [proposal](https://github.com/Expensify/App/issues/46350#issuecomment-2253721504).
